### PR TITLE
[fix]:[CDS-80150]:Bamboo failing at service step

### PR DESCRIPTION
**Pull request by Rakshitharness**
## Describe your changes
Now we are calling the get metadata for each artifactpath only not for each artifactpath and each build.
Screenshot 2023-10-19 at 1 58 35 PM
Screenshot 2023-10-19 at 2 00 13 PM

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/54315)

### CODE REVIEW

**Comment by Rakshitharness** [comment]
trigger messagecheck

**Comment by Rakshitharness** [comment]
trigger uts

**Review by rafaelmart1ns — APPROVED** [review]

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/bamboo/BambooArtifactTaskHandler.java [hunk 1]
```diff
@@ -110,6 +107,13 @@ public ArtifactTaskExecutionResponse getLastSuccessfulBuild(BambooArtifactDelega
 
       checkIfEmptyBuilds(buildDetails);
 
+      if (isNotEmpty(attributesRequest.getArtifactPaths())) {
+        List<ArtifactFileMetadata> artifactFileMetadata = bambooService.getArtifactFileMetadataList(
+            BambooRequestResponseMapper.toBambooConfig(attributesRequest), attributesRequest.getEncryptedDataDetails(),
+            attributesRequest.getPlanKey(), buildDetails.get(0).getNumber(), attributesRequest.getArtifactPaths());
```
Comment (naman-goenka) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/bamboo/BambooArtifactTaskHandler.java:113]: is `buildDetails` non-empty always?

Comment (Rakshitharness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/bamboo/BambooArtifactTaskHandler.java:113]: we check if builds are empty or not over here
checkIfEmptyBuilds(buildDetails);

File: 930-delegate-tasks/src/test/java/io/harness/delegate/task/artifacts/bamboo/BambooArtifactTaskHandlerTest.java [hunk 2]
```diff
@@ -178,9 +180,10 @@ public void getLastSuccessfulBuildTest_FixedValue() {
                                                                       .buildNumber("45")
                                                                       .build();
 
-    when(bambooBuildService.getBuildsWithoutTimeOut(ARTIFACT_STREAM_ATTRIBUTES,
-             BambooRequestResponseMapper.toBambooConfig(bambooArtifactDelegateRequest),
-             bambooArtifactDelegateRequest.getEncryptedDataDetails(), Integer.MAX_VALUE))
+    when(
```
Comment (Abhishek-CDC) [inline 930-delegate-tasks/src/test/java/io/harness/delegate/task/artifacts/bamboo/BambooArtifactTaskHandlerTest.java:183]: Please add UTs/cases for metadata check also. As we are calling a new method for it now.
There is no assertion to check that in these tests.

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/bamboo/BambooArtifactTaskHandler.java [hunk 3]
```diff
@@ -110,6 +107,13 @@ public ArtifactTaskExecutionResponse getLastSuccessfulBuild(BambooArtifactDelega
...
+        buildDetails.get(0).setArtifactDownloadMetadata(artifactFileMetadata);
```
Comment (Abhishek-CDC) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/bamboo/BambooArtifactTaskHandler.java:114]: Nit: Lets create a variable to store buildDetails.get(0) and we can perform operations on it and use in place of referring array every time

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/bamboo/BambooArtifactTaskHandler.java (hunk 1 above)
Comment (Abhishek-CDC) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/bamboo/BambooArtifactTaskHandler.java:113]: we are checking this in line 108 -> checkIfEmptyBuilds(buildDetails);

**Comment by Abhishek-CDC** [comment]
Please add screenshot of scenarios tested

**Review by abhinavmittal-harness — COMMENTED** [review]
> Reviewed 4 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54315)* status: 5 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Comment by Rakshitharness** [comment]
trigger uts

**Review by Abhishek-CDC — APPROVED** [review]

**Review by naman-goenka — CHANGES_REQUESTED** [review]
> Reviewed 1 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54315)* status: 5 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/test/java/io/harness/delegate/task/artifacts/bamboo/BambooArtifactTaskHandlerTest.java` line 209 at r2](https://reviewable.io/reviews/harness/harness-core/54315#-NhBK4je-uGVbPmIHxIh:-NhBK4je-uGVbPmIHxIi:b-977qky) ([raw file](https://github.com/harness/harness-core/blob/0b313c420e299d33fca8d8d7e1b49fdf612af7c7/930-delegate-tasks/src/test/java/io/harness/delegate/task/artifacts/bamboo/BambooArtifactTaskHandlerTest.java#L209)):*
> ```Java
>     assertThat(buildDetails.getArtifactPath()).isEqualTo("artifactPath");
>     assertThat(buildDetails.getNumber()).isEqualTo("45");
>     assertThat(buildDetails.getArtifactDownloadMetadata()).isEqualTo(Artifact_MetadataList);
> ```

Can you add test case for attributesRequest.getArtifactPaths() being empty and then new API call not being made

**Review by naman-goenka — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54315)* status: 5 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/software/wings/helpers/ext/bamboo/BambooServiceImpl.java` line 562 at r2](https://reviewable.io/reviews/harness/harness-core/54315#-NhBKuNQ-w64z3dvRd5I:-NhBKuNQ-w64z3dvRd5J:bkvkti6) ([raw file](https://github.com/harness/harness-core/blob/0b313c420e299d33fca8d8d7e1b49fdf612af7c7/930-delegate-tasks/src/main/java/software/wings/helpers/ext/bamboo/BambooServiceImpl.java#L562)):*
> ```Java
> 
>   public List<ArtifactFileMetadata> getArtifactFileMetadataList(BambooConfig bambooConfig,
>       List<EncryptedDataDetail> encryptionDetails, String planKey, String buildNumber, List<String> artifactPathRegex) {
> ```

consider adding a notNull annottion/check on artifactPathRegex

**Comment by Rakshitharness** [comment]
trigger uts

**Review by Abhishek-CDC — APPROVED** [review]

**Review by rafaelmart1ns — APPROVED** [review]

**Review by naman-goenka — APPROVED** [review]

**Comment by Rakshitharness** [comment]
trigger review

**Pull request merged**